chain/chain/src/chain.rs

```diff
@@ -2466,6 +2473,41 @@ impl Chain {
     }
 }
 
+/// Ganache node specific operations
+#[cfg(feature = "ganache")]
+impl Chain {
+    pub fn patch_state<Record: Borrow<StateRecord>>(&mut self, records: &[Record]) {
+        let mut store_update = self.store.owned_store().store_update();
+        for record in records {
+            match record.borrow().clone() {
+                StateRecord::Account { account_id, account } => store_update.set(
+                    ColState,
+                    &(TrieKey::Account { account_id }.to_vec()),
+                    &account.try_to_vec()?,
+                ),
+                StateRecord::AccessKey { account_id, public_key, access_key } => store_update.set(
+                    ColState,
+                    &(TrieKey::AccessKey { account_id, public_key }.to_vec()),
+                    &access_key.try_to_vec()?,
+                ),
+                StateRecord::Contract { account_id, code } => store_update.set(
+                    ColState,
+                    &(TrieKey::ContractCode { account_id }.to_vec()),
+                    &code,
+                ),
+                StateRecord::Data { account_id, data_key, value } => store_update.set(
+                    ColState,
+                    &(TrieKey::ContractData { account_id, key: data_key }.to_vec()),
+                    &value,
+                ),
+                _ => unimplemented!("patch_state can only patch Account, AccessKey, Contract and Data kind of StateRecord")
+            }
+        }
+        // store_update.set(ColStateParts, &key, &state_part);
+        store_update.commit().expect("Failed to patch state");
```

Comment:
I don't think this is what you want. This implementation will only change what is stored in the database but will not update trie. Also implementing this in `Chain` breaks the invariant that state update is only done in runtime. Even for features that are not enabled by default I think it is better to uphold those invariants.

Reply:
Yeah I just figured out it doesn't update trie so test didn't pass

> Also implementing this in Chain breaks the invariant that state update is only done in runtime. Even for features that are not enabled by default I think it is better to uphold those invariants.

Good point, I'm thinking of a better way to do this. By adding a ganache node only `state_patches` argument to `runtime.apply` that result in trie_update and state change. In chain just pass this argument to runtime. state_patches should not be considered as a type of transaction or receipt because it's fundamentally different. WDYT?

Reply:
I took a further look on this, maybe it's also okay to make patch_state a free, only allowed in ganache-node action in transaction. Because trie changes are applied to store on each block by `Chain::process_block_single`, so block is the single unit of work that caused a trie and state changed. Therefore state_patch should be part of a block and handled by `apply_chunks->apply_transactions->runtime.apply` that returns `{..., trie_changes, state_changes}` and resulting trie_changes commit by `Chain::process_block_single` makes patch state effective.
There're two ways to include a patch state in block: as a new field, or as a special kind of action in transactions.

Reply:
@bowenwang1996 Would like to see your thought on the potential impact of chain and keep most of the storage invariants on one of the two two approaches:

>  There're two ways to include a patch state in block: as a new field, or as a special kind of action in transactions.

Reply:
Let's figure out why we need this operation first.

Reply:
We confirmed we do need this operation because ethereum developers use this feature in writing tests and they would expect this feature. And discussed offline for an implementation plan to store patch_state on chain instead of in block. Because we should avoid adding thing to block and transaction even when it's under a feature. And when client actor got a patch state request from rpc, it only has access to chain client and chain. So it can store this info in chain, I name the field Chain.states_to_patch , then in Chain::process_block_single it create a ChainUpdate and pass the states_to_patch to ChainUpdate, and ChainUpdate can pass states_to_patch to RuntimeAdapter::apply_transactions. So I'll add Chain.states_to_patch and ChainUpdate.states_to_patch field. 
This described is implemented, WIP clean up + test.